### PR TITLE
Provide error range for src-annotation

### PR DESCRIPTION
**Pull request by o-farag**
Solves https://github.com/eclipse/lemminx/issues/1276

Also provide code action to change to `xs:appinfo` or `xs:documentation`

image
image\
image

### CODE REVIEW

**Comment by o-farag** [comment]
@datho7561 Two concerns I have with this are

1. The code action doesn't indicate what text will be replaced i.e. it only says `Replace with "appinfo"` instead of `Replace "appInfo" with "appinfo"`
2. When the code action corrects the annotation, only the opening tag is fixed. Another code action comes up later with these options
image
Ideally I'd like to have the closing tag fixed with just the original code action

These are the 2 things I'm trying to figure out how to do right now. If you think it's good enough as is right now I can open the PR for review. Otherwise I'll try to figure these 2 things out.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+			DOMElement element = (DOMElement) node;
+			String tagName = element.getTagName();
+			codeActionText = "Replace '" + tagName + "' with ";
+		} catch (BadLocationException e) {
+			codeActionText = "Replace with ";
+		}
+		
+	
+		// Replace with "appinfo"
+		CodeAction replaceAction_appinfo = CodeActionFactory.replace(codeActionText + "'xs:appinfo'", diagnosticRange,
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java:41]: You replace just the start tag name, but you should replace the end tagname (as you have noticed). To do that I suggest that you build a lits of TextEdit with those 2 replacements by using https://github.com/eclipse/lemminx/blob/b0ab7dc60f78095525376ba165cd5299b9421db3/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/CodeActionFactory.java#L108

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		try {
+			int startOffset = document.offsetAt(diagnosticRange.getStart()) + 1;
+			DOMNode node = document.findNodeAt(startOffset);
+			DOMElement element = (DOMElement) node;
+			String tagName = element.getTagName();
+			codeActionText = "Replace '" + tagName + "' with ";
+		} catch (BadLocationException e) {
+			codeActionText = "Replace with ";
+		}
+
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java:38]: remove blank line

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package org.eclipse.lemminx.extensions.contentmodel.participants.codeactions;
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java:1]: Add header

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import org.eclipse.lemminx.dom.DOMNode;
+import org.eclipse.lemminx.services.extensions.codeaction.ICodeActionParticipant;
+import org.eclipse.lemminx.services.extensions.codeaction.ICodeActionRequest;
+
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lemminx.commons.CodeActionFactory;
+
+public class src_annotationCodeAction implements ICodeActionParticipant {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java:19]: Add Javadoc which explains what does this code action

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		// Replace with "appinfo"
+		CodeAction replaceAction_appinfo = CodeActionFactory.replace(codeActionText + "'xs:appinfo'", diagnosticRange,
+				"xs:appinfo",
+				document.getTextDocument(), diagnostic);
+		
+		// Replace with "documentation"
+		CodeAction replaceAction_documentation = CodeActionFactory.replace(codeActionText + "'xs:documentation'", diagnosticRange,
+				"xs:documentation",
+				document.getTextDocument(), diagnostic);
+
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java:49]: Remove those 2 blank lines

**Comment by angelozerr** [comment]
> Ideally I'd like to have the closing tag fixed with just the original code action

I agree, see my suggestion at https://github.com/eclipse/lemminx/pull/1581#discussion_r1316413656

**Comment by angelozerr** [comment]
@o-farag please write tests like https://github.com/eclipse/lemminx/blob/b0ab7dc60f78095525376ba165cd5299b9421db3/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/xsd/XSDValidationExtensionsTest.java#L204

**Comment by o-farag** [comment]
@angelozerr I believe I made the requested changes. Let me know if anything else needs to be fixed. Thank you :)

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/xsd/XSDValidationExtensionsTest.java [hunk 6]
```diff
@@ -222,6 +222,28 @@ public void s4s_elt_invalid_content_3WithSelfClosingTag() throws BadLocationExce
 		testCodeActionsFor(xml, d, ca(d, te(2, 1, 2, 13, "")));
 	}
 
+	@Test
+	public void src_annotation_invalid_tag_() throws BadLocationException {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/xsd/XSDValidationExtensionsTest.java:226]: Please add  2 new tests

 *  with xs:documentation.
 * with xs:appInfo which have no end tag (<xs:annotation><xs:appInfo></xs:annotation>)

Comment (o-farag) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/xsd/XSDValidationExtensionsTest.java:226]: Hello, I'm not too sure what you mean. Doesn't this line test for xs:documentation? https://github.com/eclipse/lemminx/blob/5782e0cfe5790c47a30eb3592bc4cd5335d9250a/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/xsd/XSDValidationExtensionsTest.java#L240-L242

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+			int startOffset = document.offsetAt(diagnosticRange.getStart()) + 1;
+			DOMNode node = document.findNodeAt(startOffset);
+			DOMElement element = (DOMElement) node;
+			String tagName = element.getTagName();
+			codeActionText = "Replace '" + tagName + "' with ";
+			closeRange = XMLPositionUtility.selectEndTagName(element);
+		} catch (BadLocationException e) {
+			codeActionText = "Replace with ";
+		}
+		List<String> potentialTags = new ArrayList<String>() {
```
[the remark this answers is not in the document]
Comment (o-farag) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java]: Good point, thank you :)

[the remark this answers is not in the document]
Comment (o-farag) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/src_annotationCodeAction.java]: Good point, thank you :)

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/xsd/XSDValidationExtensionsTest.java (hunk 6 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/xsd/XSDValidationExtensionsTest.java:226]: Sorry I though there were too the same error for documentation, please ignore my comment.

**Review by datho7561 — APPROVED** [review]
Looks good and works well. Thanks, Omar!

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/xsd/XSDValidationExtensionsTest.java (hunk 6 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/xsd/XSDValidationExtensionsTest.java:226]: what about test with with xs:appInfo which have no end tag (<xs:annotation><xs:appInfo></xs:annotation>)?

**Pull request merged**